driver-core/src/main/com/mongodb/internal/time/TimePoint.java

```diff
@@ -17,72 +17,171 @@
+ * representing a point on a timeline. The origin of this timeline (which is not
+ * exposed) has no relation to the {@linkplain  Clock#systemUTC() system clock}.
+ * The same timeline is used by all {@link TimePoint}s within the same process.
  * <p>
  * Methods operating on a pair of {@link TimePoint}s,
  * for example, {@link #durationSince(TimePoint)}, {@link #compareTo(TimePoint)},
  * or producing a point from another one, for example, {@link #add(Duration)},
- * work correctly only if the duration between the points is not greater than {@link Long#MAX_VALUE} nanoseconds,
- * which is more than 292 years.</p>
+ * work correctly only if the duration between the points is not greater than
+ * {@link Long#MAX_VALUE} nanoseconds, which is more than 292 years.</p>
  * <p>
  * This class is not part of the public API and may be removed or changed at any time.</p>
  */
 @Immutable
-public final class TimePoint implements Comparable<TimePoint> {
-    private final long nanos;
+class TimePoint implements Comparable<TimePoint>, StartTime, Deadline {
+    @Nullable
+    private final Long nanos;
 
-    private TimePoint(final long nanos) {
+    TimePoint(@Nullable final Long nanos) {
         this.nanos = nanos;
     }
 
+    @VisibleForTesting(otherwise = PRIVATE)
+    static TimePoint at(@Nullable final Long nanos) {
+        return new TimePoint(nanos);
+    }
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    long currentNanos() {
+        return System.nanoTime();
+    }
+
     /**
      * Returns the current {@link TimePoint}.
      */
-    public static TimePoint now() {
+    static TimePoint now() {
         return at(System.nanoTime());
     }
 
-    @VisibleForTesting(otherwise = PRIVATE)
-    static TimePoint at(final long nanos) {
-        return new TimePoint(nanos);
+    /**
+     * Returns a {@link TimePoint} infinitely far in the future.
+     */
+    static TimePoint infinite() {
+        return at(null);
     }
 
     /**
-     * The {@link Duration} between this {@link TimePoint} and {@code t}.
-     * A {@linkplain Duration#isNegative() negative} {@link Duration} means that
-     * this {@link TimePoint} is {@linkplain #compareTo(TimePoint) before} {@code t}.
+     * @return true if this timepoint is infinite.
+     */
+    @Override
+    public boolean isInfinite() {
+        return nanos == null;
+    }
+
+    /**
+     * @return this TimePoint, as a deadline. Convenience for {@link StartTime}
+     */
+    @Override
+    public Deadline asDeadline() {
+        return this;
+    }
+
+    /**
+     * The number of whole time units that remain until this TimePoint
+     * {@link #hasExpired()}. This should not be used to check for expiry,
+     * but can be used to supply a remaining value, in the finest-grained
+     * TimeUnit available, to some method that may time out.
+     * This method must not be used with infinite TimePoints.
```

Comment:
This changes the behaviour of the `remaining` method.

Previously, if there were 100ns remaining, and you asked how many hours were remaining, it would return 1 hour, because returning 0 might suggest that the deadline is expired. However, this does not match the behaviour of Duration, and I do not think rounding up is correct, for the expected usage. For example, I might get the remaining time in hours because that is the granularity of some method that I will invoke, and get back 1, and wait on the method for the next hour, which is not correct. As far as "hours" go, the deadline is indeed expired, and I do not think we should undertake operations that might (because of granularity differences) exceed the timeout. The same logic applies to milliseconds, though less intuitively.

Reply:
Good catch

Reply:
I spoke with @stIncMale about this, and he explained that the existing rounding-up behaviour is generally consistent with the timeout interpretation elsewhere in Java, of "try at least this long and then time out as fast as possible". I agree this is not an incorrect interpretation, but we also agree that the appropriate way to address this particular issue is to use nanoseconds (as Java SE does), rather than rounding the value up and down to some granularity. I won't address this in this PR (that is, refactoring our existing methods). I will leave this rounding-up interpretation, since it is consistent with how "elapsed" works, and the inconsistency with Java SE is not crucial (and should ideally be addressed in a different way regardless).

Reply:
Ack